lib/SILOptimizer/Mandatory/TFDifferentiation.cpp

```diff
@@ -405,6 +412,313 @@ bool ADContext::supportsFloatingPointDifferentiation(Type type) const {
+      // Recursively collect.
+      collectVariedValues(val, variedValues, inputIndex, visited);
+    }
+  }
+}
+
+/// Recursively find all "useful" values relative to the given value.
+///
+/// NOTE: The given value will be considered useful.
+static void collectUsefulValues(SILValue value,
+                                SmallDenseSet<SILValue> &usefulValues,
+                                unsigned outputIndex) {
+  DEBUG(getDebugStream() << "USEFUL @ " << outputIndex << ":\n"
+        << value << '\n');
+  usefulValues.insert(value);
+  if (auto *def = value->getDefiningInstruction())
+    for (auto &op : def->getAllOperands())
+      collectUsefulValues(op.get(), usefulValues, outputIndex);
+}
+
+void DifferentiableActivityInfo::analyze() {
+  DEBUG(getDebugStream() << "Running activity analysis on @"
+        << function.getName() << '\n');
+  // Inputs are just function's arguments, count `n`.
+  auto paramArgs = function.getArgumentsWithoutIndirectResults();
+  for (auto valueAndIndex : enumerate(paramArgs)) {
+    inputValues.push_back(valueAndIndex.first);
+  }
+  DEBUG({
+    auto &s = getDebugStream();
+    s << "Inputs in @" << function.getName() << ":\n";
+    for (auto val : inputValues) s << val << '\n';
+  });
+  // Outputs are indirect result buffers and return values, count `m`.
+  collectAllFormalResultsInTypeOrder(function, outputValues);
+  DEBUG({
+    auto &s = getDebugStream();
+    s << "Outputs in @" << function.getName() << ":\n";
+    for (auto val : outputValues) s << val << '\n';
+  });
+  // Initialize sets to store useful values and varied values.
+  usefulValueSets.append(outputValues.size(), {});
+  variedValueSets.append(inputValues.size(), {});
+  // Mark varied values for each independent varible.
+  SmallDenseSet<SILValue> visitedVariedValues;
+  for (auto valAndIdx : enumerate(inputValues))
+    collectVariedValues(valAndIdx.first, variedValueSets[valAndIdx.second],
+                        valAndIdx.second, visitedVariedValues);
+  // Mark useful values for each dependent variable.
+  for (auto valAndIdx : enumerate(outputValues))
+    collectUsefulValues(valAndIdx.first, usefulValueSets[valAndIdx.second],
+                        valAndIdx.second);
+}
+
+bool DifferentiableActivityInfo::
+isIndependent(SILValue value, const SILReverseAutoDiffIndices &indices) const {
+  for (auto paramIdx : indices.parameters.set_bits())
+    if (inputValues[paramIdx] == value)
+      return true;
+  return false;
+}
+
+bool DifferentiableActivityInfo::
+isDependent(SILValue value, const SILReverseAutoDiffIndices &indices) const {
+  return inputValues[indices.source] == value;
+}
+
+bool DifferentiableActivityInfo::
+isVaried(SILValue value, unsigned independentVariableIndex) const {
+  auto &set = variedValueSets[independentVariableIndex];
+  return set.find(value) != set.end();
+}
+
+bool DifferentiableActivityInfo::
+isVaried(SILValue value, llvm::SmallBitVector parameterIndices) const {
```

Comment:
This should pass parameterIndices by const&.  Yes, it is usually small, but it still has a ctor and dtor.